zcash_primitives/src/transaction/mod.rs

```diff
@@ -228,39 +300,18 @@ impl Transaction {
     pub fn read<R: Read>(reader: R) -> io::Result<Self> {
         let mut reader = HashReader::new(reader);
 
-        let header = reader.read_u32::<LittleEndian>()?;
-        let overwintered = (header >> 31) == 1;
-        let version = header & 0x7FFFFFFF;
-
-        let version_group_id = if overwintered {
-            reader.read_u32::<LittleEndian>()?
-        } else {
-            0
-        };
-
-        let is_overwinter_v3 = overwintered
-            && version_group_id == OVERWINTER_VERSION_GROUP_ID
-            && version == OVERWINTER_TX_VERSION;
-        let is_sapling_v4 = overwintered
-            && version_group_id == SAPLING_VERSION_GROUP_ID
-            && version == SAPLING_TX_VERSION;
+        let version = TxVersion::read(&mut reader)?;
+        let is_overwinter_v3 = version == TxVersion::Overwinter;
+        let is_sapling_v4 = version == TxVersion::Sapling;
```

Comment:
Previously this would read transactions with invalid combinations of {`header`, `nVersionGroupId`} fields. Now it will not. This is not a consensus problem because the Rust code does not have responsibility for implementing these consensus checks. However, I'm not sure about other potential effects. Is this an intended change in behaviour?

Reply:
Interesting. Should we add a `TxVersion::Unrecognized` case?

Reply:
This wouldn't read invalid combinations previously; we explicitly checked valid combinations, and returned an error otherwise.

Reply:
And AFAICT, this code reads exactly the same amount of data from `reader` before erroring as the prior code did (the first 4 bytes, plus the next 4 bytes if the `overwintered` bit is set).